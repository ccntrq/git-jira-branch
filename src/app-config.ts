import * as Doc from '@effect/printer/Doc';
import {
  type Brand,
  Config,
  ConfigError,
  Context,
  Effect,
  Either,
  Layer,
  flow,
} from 'effect';
import {InvalidData} from 'effect/ConfigError';

import {
  type AppConfig,
  AppConfigError,
  JiraApiToken,
  JiraApiUrl,
  JiraCloudAuth,
  JiraDataCenterAuth,
  JiraKeyPrefix,
  JiraPat,
  JiraUserEmail,
} from './types';

const mapBrandOrFail =
  <T extends Brand.Brand<string>>(brand: Brand.Brand.Constructor<T>) =>
  (config: Config.Config<Brand.Brand.Unbranded<T>>) =>
    config.pipe(
      Config.mapOrFail((v) =>
        brand.either(v).pipe(Either.mapLeft(brandErrorToConfigError)),
      ),
    );

const brandErrorToConfigError = (
  e: Brand.Brand.BrandErrors,
): ConfigError.ConfigError =>
  InvalidData([], e.map((_) => _.message).join(' '));

const jiraCloudAuthConfig = Config.all({
  jiraApiToken: Config.string('JIRA_API_TOKEN').pipe(
    mapBrandOrFail(JiraApiToken),
  ),
  jiraUserEmail: Config.string('JIRA_USER_EMAIL').pipe(
    mapBrandOrFail(JiraUserEmail),
  ),
}).pipe(Config.map(JiraCloudAuth));

const jiraDataCenterAuthConfig = Config.all({
  jiraPat: Config.string('JIRA_PAT').pipe(mapBrandOrFail(JiraPat)),
}).pipe(Config.map(JiraDataCenterAuth));

const appConfig = Config.all({
  jiraApiUrl: Config.string('JIRA_API_URL').pipe(mapBrandOrFail(JiraApiUrl)),
  defaultJiraKeyPrefix: Config.option(
    Config.string('JIRA_KEY_PREFIX').pipe(mapBrandOrFail(JiraKeyPrefix)),
  ),
  jiraAuth: Config.orElse(jiraCloudAuthConfig, () => jiraDataCenterAuthConfig),
});

const collectConfigErrorMessagesReducer: ConfigError.ConfigErrorReducer<
  undefined,
  Doc.Doc<never>
> = (() => {
  const collectMessage = (
    _: undefined,
    _path: Array<string>,
    message: string,
  ): Doc.Doc<never> => Doc.text(message);

  return {
    andCase: (_, left, right) => Doc.vcat([left, right]),
    orCase: (_, left, right) =>
      Doc.vcat([Doc.indent(left, 2), Doc.text('or'), Doc.indent(right, 2)]),
    invalidDataCase: (_, path, message) =>
      Doc.text(`Expected valid input for ${path.join('.')} (${message})`),
    missingDataCase: collectMessage,
    sourceUnavailableCase: collectMessage,
    unsupportedCase: collectMessage,
  };
})();

// biome-ignore lint/complexity/noStaticOnlyClass: okay for AppConfigService
export class AppConfigService extends Context.Tag('AppConfigService')<
  AppConfigService,
  {
    readonly getAppConfig: Effect.Effect<AppConfig, AppConfigError, never>;
  }
>() {
  public static readonly Live = Layer.succeed(
    AppConfigService,
    AppConfigService.of({
      getAppConfig: appConfig.pipe(
        Effect.mapError(
          flow(
            ConfigError.reduceWithContext(
              void 0,
              collectConfigErrorMessagesReducer,
            ),
            Doc.render({style: 'pretty'}),
            (message) => AppConfigError({message}),
          ),
        ),
      ),
    }),
  );
}
