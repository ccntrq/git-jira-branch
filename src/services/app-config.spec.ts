import {live} from '@effect/vitest';
import {ConfigProvider, Effect, Either, Layer, pipe} from 'effect';
import * as Option from 'effect/Option';
import {describe, expect} from 'vitest';
import {
  JiraApiToken,
  JiraCloudAuth,
  JiraDataCenterAuth,
  JiraPat,
  JiraUserEmail,
} from '../types.js';
import {AppConfigService} from './app-config.js';

const testProg = AppConfigService.pipe(Effect.flatMap((_) => _.getAppConfig));

const mkTestLayer = (
  configProvider: ConfigProvider.ConfigProvider,
): Layer.Layer<AppConfigService> =>
  AppConfigService.Live.pipe(
    Layer.provide(Layer.setConfigProvider(configProvider)),
  );

describe('AppConfigService', () => {
  live('should provide jira data center appconfig', () =>
    Effect.gen(function* () {
      const configProvider = ConfigProvider.fromMap(
        new Map([
          ['JIRA_PAT', 'dummy-jira-pat'],
          ['JIRA_API_URL', 'https://dummy-jira-instance.com'],
          ['JIRA_KEY_PREFIX', 'DUMMYAPP'],
        ]),
      );

      const env = yield* Effect.provide(testProg, mkTestLayer(configProvider));

      expect(env).toEqual({
        defaultJiraKeyPrefix: Option.some('DUMMYAPP'),
        jiraApiUrl: 'https://dummy-jira-instance.com',
        jiraAuth: JiraDataCenterAuth({
          jiraPat: JiraPat('dummy-jira-pat'),
        }),
      });
    }),
  );

  live('should provide jira cloud appconfig', () =>
    Effect.gen(function* () {
      const configProvider = ConfigProvider.fromMap(
        new Map([
          ['JIRA_API_TOKEN', 'dummy-jira-api-token'],
          ['JIRA_USER_EMAIL', 'dummyuser@example.com'],
          ['JIRA_API_URL', 'https://dummy-jira-instance.com'],
          ['JIRA_KEY_PREFIX', 'DUMMYAPP'],
        ]),
      );

      const env = yield* Effect.provide(testProg, mkTestLayer(configProvider));

      expect(env).toEqual({
        defaultJiraKeyPrefix: Option.some('DUMMYAPP'),
        jiraApiUrl: 'https://dummy-jira-instance.com',
        jiraAuth: JiraCloudAuth({
          jiraApiToken: JiraApiToken('dummy-jira-api-token'),
          jiraUserEmail: JiraUserEmail('dummyuser@example.com'),
        }),
      });
    }),
  );

  live('should provide appconfig with missing optional values', () =>
    Effect.gen(function* () {
      const configProvider = ConfigProvider.fromMap(
        new Map([
          ['JIRA_PAT', 'dummy-jira-pat'],
          ['JIRA_API_URL', 'https://dummy-jira-instance.com'],
        ]),
      );

      const env = yield* Effect.provide(testProg, mkTestLayer(configProvider));

      expect(env).toEqual({
        defaultJiraKeyPrefix: Option.none(),
        jiraApiUrl: 'https://dummy-jira-instance.com',
        jiraAuth: JiraDataCenterAuth({
          jiraPat: JiraPat('dummy-jira-pat'),
        }),
      });
    }),
  );

  live('should report missing configuration values', () =>
    Effect.gen(function* () {
      const configProvider = ConfigProvider.fromMap(
        new Map([['JIRA_API_URL', 'https://dummy-jira-instance.com']]),
      );

      const result = yield* Effect.either(
        Effect.provide(testProg, mkTestLayer(configProvider)),
      );
      Either.match(result, {
        onLeft: (e) => expect(e).toMatchSnapshot(),
        onRight: (_) => expect.unreachable('Should have returned an error'),
      });
    }),
  );

  live(
    '`defaultJiraKeyPrefix` should work with missing configuration values',
    () =>
      Effect.gen(function* () {
        const configProvider = ConfigProvider.fromMap(new Map());

        return yield* Effect.provide(
          AppConfigService.pipe(Effect.flatMap((_) => _.defaultJiraKeyPrefix)),
          mkTestLayer(configProvider),
        ).pipe(
          Effect.match({
            onFailure: () => expect.unreachable('should work without config'),
            onSuccess: (d) =>
              expect(d).toMatchInlineSnapshot(`
              {
                "_id": "Option",
                "_tag": "None",
              }
            `),
          }),
        );
      }),
  );

  live('should validate config values', () =>
    Effect.gen(function* () {
      const configProvider = ConfigProvider.fromMap(
        new Map([
          ['JIRA_API_TOKEN', ''],
          ['JIRA_USER_EMAIL', ''],
          ['JIRA_API_URL', ''],
          ['JIRA_KEY_PREFIX', ''],
        ]),
      );

      yield* pipe(
        Effect.provide(testProg, mkTestLayer(configProvider)),
        Effect.match({
          onSuccess: () => expect.unreachable('Should have returned an error'),
          onFailure: (e) => expect(e).toMatchSnapshot(),
        }),
      );
    }),
  );
});
