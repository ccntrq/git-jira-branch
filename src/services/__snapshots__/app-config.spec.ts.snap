// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AppConfigService > should report missing configuration values 1`] = `
{
  "_tag": "AppConfigError",
  "message": "  Expected JIRA_API_TOKEN to exist in the provided map
  Expected JIRA_USER_EMAIL to exist in the provided map
or
  Expected JIRA_PAT to exist in the provided map",
}
`;

exports[`AppConfigService > should validate config values 1`] = `
{
  "_tag": "AppConfigError",
  "message": "Expected valid input for JIRA_API_URL ('' is not a valid Jira API URL)
Expected valid input for JIRA_KEY_PREFIX ('' is not a valid Jira key prefix)
  Expected valid input for JIRA_API_TOKEN (Provided value is not a valid Jira API token)
  Expected valid input for JIRA_USER_EMAIL ('' is not a valid email address)
or
  Expected JIRA_PAT to exist in the provided map",
}
`;
